stdlib/mexpr/cfa.mc

```diff
@@ -260,11 +270,21 @@ lang DirectConstraint = CFA
   -- lhs ⊆ rhs
   | CstrDirect { lhs: Name, rhs: Name }
 
+  sem isDirect =
+  | _ /- : AbsVal -/ -> true
+
+  sem directTransition (graph: CFAGraph) (rhs: Name) =
+  | av -> av
+
   sem initConstraint (graph: CFAGraph) =
   | CstrDirect r & cstr -> initConstraintName r.lhs graph cstr
 
   sem propagateConstraint (update: (Name,AbsVal)) (graph: CFAGraph) =
-  | CstrDirect r -> addData graph update.1 r.rhs
+  | CstrDirect r ->
+    if isDirect update.1 then
+      let av = directTransition graph r.rhs update.1 in
+      addData graph av r.rhs
+    else graph
```

Comment:
I may be wrong, but I believe you need to do an equivalent change to this part of `cfa.mc`:
```
lang NamedPatCFA = MatchCFA + NamedPat
  sem propagateMatchConstraint (graph: CFAGraph) (id: Name) =
  | (PatNamed { ident = PName n }, av) -> addData graph av n

```

Reply:
Yes, you're right, thanks! I put the logic in a separate language fragment `DirectConstraintOptions` that is now used both by `DirectConstraint` and `NamedPatCFA`.